L1Trigger/Phase2L1ParticleFlow/plugins/L1TPFProducer.cc

```diff
@@ -22,254 +24,375 @@
+			const l1t::L1TkMuonParticle & mu = *it;
+			if (debugR_ > 0 && deltaR(mu.eta(),mu.phi(),debugEta_,debugPhi_) > debugR_) continue;
+			l1regions_.addMuon(mu);  // FIXME add a l1t::PFCandidate::MuonRef
+		}
+	}
+
+	// ------ READ CALOS -----
+	edm::Handle<l1t::PFClusterCollection> caloHandle;
+	for (const auto & tag : emCands_) {
+		iEvent.getByToken(tag, caloHandle);
+		const auto & calos = *caloHandle;
+		for (unsigned int ic = 0, nc = calos.size(); ic < nc; ++ic) {
+			const auto & calo = calos[ic];
+			if (debugR_ > 0 && deltaR(calo.eta(),calo.phi(),debugEta_,debugPhi_) > debugR_) continue;
+			if (calo.pt() > emPtCut_) l1regions_.addEmCalo(calo, l1t::PFClusterRef(caloHandle,ic));
+		}
+	}
+	for (const auto & tag : hadCands_) {
+		iEvent.getByToken(tag, caloHandle);
+		const auto & calos = *caloHandle;
+		for (unsigned int ic = 0, nc = calos.size(); ic < nc; ++ic) {
+			const auto & calo = calos[ic];
+			if (debugR_ > 0 && deltaR(calo.eta(),calo.phi(),debugEta_,debugPhi_) > debugR_) continue;
+			if (calo.pt() > hadPtCut_) l1regions_.addCalo(calo, l1t::PFClusterRef(caloHandle,ic));
+		}
+	}
+
+	// First, get a copy of the discretized and corrected inputs, and write them out
+	// FIXME: to be implemented
+	iEvent.put(l1regions_.fetchCalo(/*ptmin=*/0.1, /*em=*/true),  "EmCalo");
+	iEvent.put(l1regions_.fetchCalo(/*ptmin=*/0.1, /*em=*/false), "Calo");
+	iEvent.put(l1regions_.fetchTracks(/*ptmin=*/0.0, /*fromPV=*/false), "TK");
+	if (fRegionDump) {
+		uint32_t run = iEvent.id().run(), lumi = iEvent.id().luminosityBlock(); uint64_t event = iEvent.id().event();
+		fwrite(&run, sizeof(uint32_t), 1, fRegionDump);
+		fwrite(&lumi, sizeof(uint32_t), 1, fRegionDump);
+		fwrite(&event, sizeof(uint64_t), 1, fRegionDump);
+		l1tpf_impl::writeManyToFile(l1regions_.regions(), fRegionDump);
+	}
+
+	// Then save the regions to the COE file
+	// Do it here because there is some sorting going on in a later function
+	if (fRegionCOE && neventsproduced<neventscoemax) {
+		//auto t = l1regions_.fetchTracks()->at(0);
+		//auto tr = t.pfTrack()->track();
+		//tr->setTrackWordBits();
+		//tr->testTrackWordBits();
+
+		std::vector<l1tpf_impl::Region> regions = l1regions_.regions();
+		l1tpf_impl::PropagatedTrack current_track;
+		ap_uint<tracksize> current_track_word;
+		std::string bset_string_ = "";
+		bool has_track = false, current_bit = false;
+		for (unsigned int irow = 0; irow < ntracksmax; irow++) {
+			for (unsigned int iphi = 0; iphi < regions.size(); iphi++) {
+				if (regions[iphi].track.size()<=irow) has_track = false;
+				else has_track = true;
+
+				if (has_track) {
+					current_track = regions[iphi].track[irow];
+					current_track_word = (ap_uint<14>(0), ap_uint<7>(current_track.hwStubs), ap_uint<3>(0), ap_uint<4>(current_track.hwChi2),
+										  ap_int<13>(0), ap_int<12>(current_track.hwZ0), ap_int<16>(current_track.hwVtxEta), ap_int<12>(current_track.hwVtxPhi),
+										  current_track.hwCharge,ap_uint<14>(current_track.hwPt));
```

Comment:
in the raw track word, it's it stored as 1/pt ?

Reply:
also, I had understood the raw track word was now accessible from the track object itself.

Reply:
The updated TTTrack, and hence the TTTrackWord, is not yet available. When that happens, I will update this code to use the official track word. When that happens, the pT will be updated to be q/R (the actual quantity stored) and eta will be updated to be tan(lambda). I don't currently have access to R or to lambda, so I chose to keep pT and eta for now.